### ISSUE TITLE
No Args Constructor question

### ISSUE THREAD

**Issue by lechiefe** [javacc21/javacc21#244]
Hi, I am very new to the library,  I am attempting something that should be fairly simple, however I am not able to find any working examples, in the /examples directory or elsewhere online. Could someone explain to me how to create a parser that has a NoArgs Constructor and instead a simple parse(String input) method? So the Api I am looking to generate would look something like this: 
`public void doSomething() {
        MyParser parser = new MyParser();
        Query result = parser.parse("Hello World");
}`

In other words I would like to be able to easily reuse the same object by calling the parse method.

Thanks in advance!

**Comment by revusky** [comment]
> In other words I would like to be able to easily reuse the same object by calling the parse method.

Well, the overall design of the thing is that the XXXParser (and XXXLexer) objects are single-use objects. This has certain key advantages, by the way -- particularly in terms of thinking about thread-safety...

However, I'm not sure what your practical problem is. If you just want to have an API that looks like: `parse(someInput)` where input is just a string, you could perfectly well just *inject* the appropriate method. So, for example, right within your grammar, you could have something like:

      INJECT PARSER_CLASS :
      {
           static public Query parse(String input) {
                  MyParser parser = new MyParser(input);
                  return parser.Query();
           }
      }

Then your client code could just be:

         Query result = MyParser.parse("Hello, World");

Of course, you're not reusing the same MyParser object. It's instantiating a new one behind the scenes each time, and you're just putting a facade in front, but I can't really see why that is a problem?!

**Comment by lechiefe** [comment]
Hi, thanks for the quick response. That's not quite what I am looking for, ideally I could just flush the buffer and not create a new object. The main concern is not creating as much gc pressure, but it sounds like that is not possible here. The application I was looking to use it for would potentially have a very high throughput of small queries, so creating a new object for each one is unwanted.

**Comment by revusky** [comment]
> ... so creating a new object for each one is unwanted.

Well, frankly, I think your concern is misplaced. The fact is that, behind the scenes, any parsing operation is creating quite a few objects behind the scenes. Like, for example, consider something like:

       x = (y + z)/2

That line, on its own is a valid python source file. So you could try this:

        git clone https://github.com/javacc21/javacc21.git
        cd javacc21/examples/python
        ant compile
        java PyTest trivial.py

I'm assuming, by the way, that you have a file called `trivial.py` with that one line in it. See what that outputs. It dumps the parse tree:

    <Module (1, 1)-(1, 15)>
      <Statement (1, 1)-(1, 15)>
        <Assignment (1, 1)-(1, 14)>
          x
          =
          <MultiplicativeExpression (1, 5)-(1, 14)>
            <Group (1, 5)-(1, 11)>
              (
              <AdditiveExpression (1, 6)-(1, 10)>
                y
                +
                z
              )
            /
           7
      EOF

That's a flat representation of the tree that is created when you parse that single line. So you see that even parsing  something that trivial involves quite a bit of object creation! It's creating tokens and non-terminal nodes, and the PythonLexer object contains a few tables that track where the tokens are in the input and so on...

Even if you could somehow reuse the outer `PythonParser object`, it is very very dubious that this could make any noticeable difference in terms of stressing the garbage collector, because it would still be creating all these other objects every time you feed it some input to parse!

That each invocation of the machinery creates a lot of objects that later need to be garbage-collected is just in the nature of things. I think you will eventually realize that it's just not worth worrying about.

**Comment by lechiefe** [comment]
Thank you for the detailed explanation. I see your point, although for small samples, like the one you show above it seems that the syntax tree is much more lightweight than the very large objects that are needed to compute it. But I agree that it will still probably not create a meaningful difference.

**Comment by lechiefe** [comment]
As long as I already have an issue open, I will ask it here, but if let me know if you want me to create a separate issue for this. 
I am assuming this is user error, but I am having a really strange issue with the code generating an off-by-one error. 

The following definition: 
```
<Boost> TOKEN : {

  <NUMBER:         (<_NUM_CHAR>)+ ( "." (<_NUM_CHAR>)+ )? > : DEFAULT
}
```

produces the following code: 

```static private void NFA_FUNCTIONS_init() { 

            NfaFunction[] functions= new NfaFunction[4]; 

            functions[1]= Boost::NFA_COMPOSITE_Boost_1;

            functions[2]= Boost::NFA_Boost_2;

            functions[3]= Boost::NFA_Boost_3;

            functions[4]= Boost::NFA_Boost_4;

            functionTableMap.put(LexicalState.Boost, functions);

        }

``` 
As you can see that will end up throwing an IndexOutOfBoundException at runtime. This seems to be the only definition causing this.
Any help is appreciated. 
Thanks!

**Comment by revusky** [comment]
Well, that's definitely strange, but this is a fragment, right? Where is the rest of the grammar. Where is _NUM_CHAR defined? I have to think this is a very minimal example. Why don't you post the whole thing?

**Comment by lechiefe** [comment]
Sure, this is the rest of the grammar: 

```
<*> TOKEN : {
  <#_NUM_CHAR:        ["0"-"9"] >
| <#_ESCAPED_CHAR:    "\\" ~[] >  // every character that follows a backslash is considered as an escaped character
| <#_TERM_START_CHAR: ( ~[ " ", "\t", "\n", "\r", "\u3000", "+", "-", "!", "(", ")", ":", "^",
                           "[", "]", "\"", "{", "}", "~", "*", "?", "\\", "/" ]
                        | <_ESCAPED_CHAR> ) >
| <#_TERM_CHAR:       ( <_TERM_START_CHAR> | "-" | "+" ) >
| <#_WHITESPACE:      ( " " | "\t" | "\n" | "\r" | "\u3000") >
| <#_QUOTED_CHAR:     ( ~[ "\"", "\\" ] | <_ESCAPED_CHAR> ) >
}

<DEFAULT, Range> SKIP : {
  < <_WHITESPACE>>
}

<DEFAULT> TOKEN : {
  <AND:           ("AND" | "&&") >
| <OR:            ("OR" | "||") >
| <NOT:           ("NOT" | "!") >
| <PLUS:          "+" >
| <MINUS:         "-" >
| <BAREOPER:      ("+"|"-"|"!") <_WHITESPACE> >
| <LPAREN:        "(" >
| <RPAREN:        ")" >
| <COLON:         ":" >
| <STAR:          "*" >
| <CARAT:         "^" > : Boost
| <QUOTED:        "\"" (<_QUOTED_CHAR>)* "\"">
| <TERM:          <_TERM_START_CHAR> (<_TERM_CHAR>)*  >
| <FUZZY_SLOP:    "~" ((<_NUM_CHAR>)+ (( "." (<_NUM_CHAR>)+ )? (<_TERM_CHAR>)*) | (<_TERM_CHAR>)*) >
| <PREFIXTERM:    ("*") | ( <_TERM_START_CHAR> (<_TERM_CHAR>)* "*" ) >
| <WILDTERM:      (<_TERM_START_CHAR> | [ "*", "?" ]) (<_TERM_CHAR> | ( [ "*", "?" ] ))* >
| <REGEXPTERM:    "/" (~[ "/" ] | "\\/" )* "/" >
| <RANGEIN_START: "[" > : Range
| <RANGEEX_START: "{" > : Range
}

<Boost> TOKEN : {
  <NUMBER:         <_NUM_CHAR> ( "." (<_NUM_CHAR>)+ )? > : DEFAULT
}

<Range> TOKEN : {
  <RANGE_TO:     "TO">
| <RANGEIN_END:  "]"> : DEFAULT
| <RANGEEX_END:  "}"> : DEFAULT
| <RANGE_QUOTED: "\"" (~["\""] | "\\\"")+ "\"">
| <RANGE_GOOP:   (~[ " ", "]", "}" ])+ >
}

**Comment by lechiefe** [comment]
I apologize, I actually sent the code that is working, the one that causes the error would have to have this instead on: 
```
<Boost> TOKEN : {
  <NUMBER: (<_NUM_CHAR>)+ ( "." (<_NUM_CHAR>)+ )? > : DEFAULT
}
```
I will edit the original to reflect that

**Comment by revusky** [comment]
Actually, I reduced it to the most minimal case where the bug appears:

      TOKEN : {
         <AB : ("a")+ "b">
      }

It's a very strange bug because I think, for one thing, it's only possible if there is exactly one token type in the lexical state. But [the fix](https://github.com/javacc21/javacc21/commit/91e9767268a7302254a84a5afbd3f82914891d9b) is on a single line actually. But anyway, if you grab a newer version, I think you'll see that this bug is gone.

Thanks so much for the help. Oh, by the way, feel free to join us on the the [discussion forum I've set up](https://discuss.congo.org/).

**Comment by lechiefe** [comment]
Oh wow, that was fast, thank you so much! Glad it was an easy fix. I will definitely check out the forum, thanks again for the help! I will close this issue, since its been resolved now.
